packages/x-data-grid/src/components/containers/GridRootStyles.ts

```diff
@@ -564,6 +564,7 @@ export const GridRootStyles = styled('div', {
       lineHeight: 'inherit',
     },
     [`& .${c.cellEmpty}`]: {
+      flex: 1,
```

Comment:
This way the empty cell always takes up the right amount of space in a row. Shrinks when the row is full of cells, grows when there aren't enough to fill a row.